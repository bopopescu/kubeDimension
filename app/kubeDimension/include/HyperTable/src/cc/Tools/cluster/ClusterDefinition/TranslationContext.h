/* -*- c++ -*-
 * Copyright (C) 2007-2015 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 3 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/// @file
/// Declarations for TranslationContext.
/// This file contains type declarations for TranslationContext, a class to hold
/// context populated and used by cluster definition entity translators.

#ifndef Tools_cluster_TranslationContext_h
#define Tools_cluster_TranslationContext_h

#include <map>
#include <set>
#include <string>

namespace Hypertable { namespace ClusterDefinition {

  using namespace std;

  /// @addtogroup ClusterDefinition
  /// @{

  /// Context used to translate cluster definition statements.
  class TranslationContext {

  public:

    /// Map of variable names to default values
    map<string, string> symbols;

    /// Set of role names
    set<string> roles;

    /// Map of tasks to descriptive text
    map<string, string> tasks;

    // Map of tasks to comma separated roles
    map<string, string> task_roles;
  };

  /// @}

}}

#endif // Tools_cluster_TranslationContext_h
