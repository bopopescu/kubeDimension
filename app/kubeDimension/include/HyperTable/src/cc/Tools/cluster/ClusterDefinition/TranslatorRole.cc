/*
 * Copyright (C) 2007-2015 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 3 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/// @file
/// Definitions for TranslatorRole.
/// This file contains type definitions for TranslatorRole, a class for
/// translating a role definition statement.

#include <Common/Compat.h>

#include "TranslatorRole.h"

#include <Common/Error.h>
#include <Common/Logger.h>
#include <Common/String.h>

#include <boost/algorithm/string.hpp>

#include <cctype>

using namespace Hypertable;
using namespace Hypertable::ClusterDefinition;
using namespace std;

const string TranslatorRole::translate(TranslationContext &context) {
  string translated_text;
  string translated_value;
  size_t offset = m_text.find_first_of("role:");
  if (offset == string::npos)
    HT_THROWF(Error::SYNTAX_ERROR, "Bad role definition on line %d of '%s'",
              (int)m_lineno, m_fname.c_str());
  const char *ptr = m_text.c_str() + offset + 5;
  // skip whitespace
  while (*ptr && isspace(*ptr))
    ptr++;

  const char *base = ptr;
  if (*ptr == 0 || !isalpha(*ptr))
    HT_THROWF(Error::SYNTAX_ERROR, "Bad role definition on line %d of '%s'",
              (int)m_lineno, m_fname.c_str());
  while (*ptr && (isalnum(*ptr) || *ptr == '_'))
    ptr++;

  string name(base, ptr-base);

  if (name.compare("all") == 0)
    HT_THROWF(Error::SYNTAX_ERROR, "Invalid role name 'all' on line %d of '%s'",
              (int)m_lineno, m_fname.c_str());

  if (*ptr == 0 || !isspace(*ptr))
    HT_THROWF(Error::SYNTAX_ERROR,
              "Invalid role specification for '%s' on line %d of '%s'",
              name.c_str(), (int)m_lineno, m_fname.c_str());

  while (*ptr && isspace(*ptr))
    ptr++;

  base = 0;
  while (*ptr) {
    if (base == 0 && *ptr == '-') {
      translated_value.append(1, *ptr);
    }
    else if (isalnum(*ptr) || *ptr == '_' || *ptr == '-' || *ptr == '.') {
      if (base == 0)
        base = ptr;
    }
    else if (isspace(*ptr) || *ptr == '+' ||
             *ptr == ',' || *ptr == '(' || *ptr == ')') {
      if (base) {
        string name(base, ptr-base);
        if (context.roles.count(name) > 0) {
          translated_value.append("${ROLE_");
          translated_value.append(name);
          translated_value.append("}");
        }
        else
          translated_value.append(name);
        base = 0;
      }
      if (isspace(*ptr)) {
        if (!isspace(translated_value[translated_value.length()-1]))
          translated_value.append(1, ' ');
      }
      else
        translated_value.append(1, *ptr);
    }
    else if (*ptr == '[' || *ptr == ']') {
      if (base) {
        translated_value.append(base, ptr-base);
        base = 0;
      }
      translated_value.append(1, *ptr);      
    }
    else
      HT_THROWF(Error::SYNTAX_ERROR, "Invalid character '%c' found in role "
                "definition on line %d of '%s'",
                *ptr, (int)m_lineno, m_fname.c_str());
    ptr++;
  }

  boost::trim_right_if(translated_value, boost::is_any_of(" \t\n\r"));

  translated_text.append("ROLE_");
  translated_text.append(name);
  translated_text.append("=\"");
  translated_text.append(translated_value);
  translated_text.append("\"\n");

  context.roles.insert(name);
  context.symbols["ROLE_"+name] = translated_value;

  return translated_text;
}

