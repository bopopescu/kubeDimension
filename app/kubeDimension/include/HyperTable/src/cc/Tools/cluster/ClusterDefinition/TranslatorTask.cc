/*
 * Copyright (C) 2007-2015 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 3 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/// @file
/// Definitions for TranslatorTask.
/// This file contains type definitions for TranslatorTask, a class for
/// translating a task definition statement.

#include <Common/Compat.h>

#include "TokenizerTools.h"
#include "TranslatorTask.h"

#include <Common/Error.h>
#include <Common/Logger.h>
#include <Common/String.h>
#include <Common/System.h>

#include <boost/algorithm/string.hpp>
#include <boost/tokenizer.hpp>

#include <cctype>
#include <vector>

using namespace Hypertable;
using namespace Hypertable::ClusterDefinition;
using namespace boost;
using namespace std;

namespace {

  const char *builtin_task_name[] = {
    "CLUSTER_BUILTIN_display_line",
    "show_variables",
    "with",
    nullptr
  };

  const string determine_trailing_indentation(const string &text) {
    string indentation;
    if (!text.empty()) {
      const char *base = text.c_str();
      const char *ptr = base + (text.length() - 1);
      size_t count = 0;
      while (ptr >= base && isspace(*ptr) && *ptr != '\n' && *ptr != '\r') {
        count++;
        ptr--;
      }
      indentation.append(ptr+1);
    }
    // If can't determine indentation, make it two spaces
    if (indentation.empty())
      indentation.append("  ");
    return indentation;
  }

  bool translate_ssh_statement(const char *base, const char *end,
                               const string &indentation,
                               const char **nextp, string &ssh_command,
                               string &errmsg) {
    HT_ASSERT(strncmp(base, "ssh:", 4) == 0);
    errmsg.clear();

    base += 4;

    const char *open_curly = base;
    while (open_curly < end && *open_curly != '{')
      open_curly++;

    const char *close_curly;
    if (open_curly == end || !TokenizerTools::find_end_char(open_curly, &close_curly)) {
      errmsg.append("curly brace not found");
      return false;
    }
        
    // parse options
    string options;
    {
      string text(base, open_curly-base);
      char_separator<char> sep(" \t\n\r");
      tokenizer<char_separator<char>> tokens(text, sep);
      for (const auto& token : tokens) {
        if (!strncmp(token.c_str(), "random-start-delay=", 19)) {
          if (!TokenizerTools::is_number(token.substr(19))) {
            errmsg.append("invalid random-start-delay argument");
            return false;
          }
          options.append(" --");
          options.append(token.substr(0, 18));
          options.append(" ");
          options.append(token.substr(19));
        }
        else if (token.compare("in-series") == 0) {
          options.append(" --");
          options.append(token);
        }
        else {
          errmsg.append(Hypertable::format("Unknown option '%s'", token.c_str()));
          return false;
        }
      }
    }

    // build ssh command
    ssh_command.clear();
    ssh_command.append(Hypertable::format("%s/bin/ht ssh%s \" ${_SSH_HOSTS}\" \"",
                                          System::install_dir.c_str(), options.c_str()));
    open_curly++;
    string content(open_curly, close_curly-open_curly);
    trim(content);

    string escaped_content;
    escaped_content.reserve(content.length()+32);
    for (const char *ptr = content.c_str(); *ptr; ptr++) {
      if (*ptr == '"')
        escaped_content.append(1, '\\');
      escaped_content.append(1, *ptr);
    }

    ssh_command.append(escaped_content);
    ssh_command.append("\"\n");
    ssh_command.append(indentation);
    ssh_command.append("if [ $? -ne 0 ]; then\n");
    ssh_command.append(indentation);
    ssh_command.append("  exit 1\n");
    ssh_command.append(indentation);
    ssh_command.append("fi");

    *nextp = close_curly + 1;

    return true;
  }

}




const string TranslatorTask::translate(TranslationContext &context) {
  string translated_text;
  string description;

  const char *base = m_text.c_str();
  const char *end;
  string line;

  while (*base) {

    end = strchr(base, '\n');
    line.clear();
    if (end == 0)
      line.append(base);
    else
      line.append(base, end-base);
    boost::trim(line);

    if (line[0] == '#') {
      line = line.substr(1);
      trim(line);
      if (description.empty())
        description.append(line);
      else if (description[description.length()-1] == '.') {
        description.append("  ");
        description.append(line);
      }
      else {
        description.append(" ");
        description.append(line);
      }
    }
    else
      break;
    if (end)
      base = end;
    else
      base += strlen(base);
    if (*base)
      base++;
  }

  // Do variable substitution
  while (TokenizerTools::substitute_variables(description, description,
                                              context.symbols))
    ;

  if (*base == 0)
    HT_THROWF(Error::SYNTAX_ERROR, "Bad task definition on line %d of '%s'",
              (int)m_lineno, m_fname.c_str());

  const char *ptr;
  if ((ptr = strchr(base, '{')) == 0)
    HT_THROWF(Error::SYNTAX_ERROR, "Bad task definition on line %d of '%s'",
              (int)m_lineno, m_fname.c_str());

  if (!TokenizerTools::find_end_char(ptr, &end))
    HT_THROWF(Error::SYNTAX_ERROR,
              "Missing terminating '}' in task definition on line %d of '%s'",
              (int)m_lineno, m_fname.c_str());

  string text(base, ptr-base);

  vector<string> words;
  {
    char_separator<char> sep(" ");
    tokenizer<char_separator<char>> tokens(text, sep);
    for (const auto& t : tokens)
      words.push_back(t);
  }

  if (words.size() < 2 || words.size() == 3 || words[0].compare("task:") != 0 ||
      (words.size() >= 3 && words[2].compare("roles:") != 0))      
    HT_THROWF(Error::SYNTAX_ERROR, "Bad task definition on line %d of '%s'",
              (int)m_lineno, m_fname.c_str());

  string task_name(words[1]);

  for (size_t i=0; builtin_task_name[i]; i++) {
    if (task_name.compare(builtin_task_name[i]) == 0)
      HT_THROWF(Error::SYNTAX_ERROR,
                "Task name '%s' conflicts with built-in on line %d of '%s'",
                task_name.c_str(), (int)m_lineno, m_fname.c_str());
  }

  if (!TokenizerTools::is_valid_identifier(task_name))
    HT_THROWF(Error::SYNTAX_ERROR, "Invalid task name (%s) on line %d of '%s'",
              task_name.c_str(), (int)m_lineno, m_fname.c_str());

  translated_text.append(task_name);
  translated_text.append(" () {\n  local _SSH_HOSTS=\"");;

  string task_roles;
  if (words.size() > 3) {
    text.clear();
    for (size_t i=3; i<words.size(); i++) {
      text.append(words[i]);
      text.append(" ");
    }
    char_separator<char> sep(" ,");
    tokenizer<char_separator<char>> tokens(text, sep);
    bool first = true;
    for (const auto& t : tokens) {
      if (context.roles.count(t) == 0)
        HT_THROWF(Error::SYNTAX_ERROR, "Undefined role (%s) on line %d of '%s'",
                  t.c_str(), (int)m_lineno, m_fname.c_str());
      if (first)
        first = false;
      else
        translated_text.append(" + ");
      translated_text.append("(${ROLE_");
      translated_text.append(t);
      translated_text.append("})");
      if (!task_roles.empty())
        task_roles.append(", ");
      task_roles.append(t);
    }
  }
  else {
    bool first = true;
    for (auto & role : context.roles) {
      if (first)
        first = false;
      else
        translated_text.append(" + ");
      translated_text.append("(${ROLE_");
      translated_text.append(role);
      translated_text.append("})");
    }
  }
  translated_text.append("\"\n");
  translated_text.append("  if [ $# -gt 0 ] && [ $1 == \"on\" ]; then\n");
  translated_text.append("    shift\n");
  translated_text.append("    if [ $# -eq 0 ]; then\n");
  translated_text.append("      echo \"Missing host specification in 'on' argument\"\n");
  translated_text.append("      exit 1\n");
  translated_text.append("    else\n");
  translated_text.append("      _SSH_HOSTS=\"$1\"\n");
  translated_text.append("      shift\n");
  translated_text.append("    fi\n");
  translated_text.append("  fi\n");
  translated_text.append("  echo \"");
  translated_text.append(task_name);
  translated_text.append(" $@\"\n  ");

  size_t lineno = m_lineno + TokenizerTools::count_newlines(m_text.c_str(), base);

  // skip '{' and whitespace
  base = ptr + 1;
  while (base < end && isspace(*base)) {
    if (*base == '\n')
      lineno++;
    base++;
  }

  HT_ASSERT(base <= end);

  string task_body;
  string ssh_command;
  string error_msg;
  size_t offset;

  // Check for missing ':' character
  if (TokenizerTools::find_token("ssh", base, end, &offset)) {
    size_t tmp_lineno = lineno + TokenizerTools::count_newlines(base, base+offset);
    ptr = base + offset + 3;
    while (*ptr && isspace(*ptr))
      ptr++;
    if (*ptr == '{' || *ptr == 0 || !strncmp(ptr, "random-start-delay=", 19) ||
        !strncmp(ptr, "in-series", 9))
      HT_THROWF(Error::SYNTAX_ERROR,"Invalid ssh: statement on line %d of '%s'",
                (int)tmp_lineno, m_fname.c_str());
  }

  while (TokenizerTools::find_token("ssh:", base, end, &offset)) {
    lineno += TokenizerTools::count_newlines(base, base+offset);
    task_body.append(base, offset);
    base += offset;
    string indentation = determine_trailing_indentation(task_body);
    if (!translate_ssh_statement(base, end, indentation, &ptr, ssh_command, error_msg))
      HT_THROWF(Error::SYNTAX_ERROR,"Invalid ssh: statement (%s) on line %d of '%s'",
                error_msg.c_str(), (int)lineno, m_fname.c_str());
    lineno += TokenizerTools::count_newlines(base, ptr);
    task_body.append(ssh_command);
    base = ptr;
  }

  if (base < end) {
    string content(base, end-base);
    trim_right(content);
    task_body.append(content);
  }

  translated_text.append(task_body);

  translated_text.append("\n}\n");

  context.tasks[task_name] = description;
  context.task_roles[task_name] = task_roles;

  return translated_text;
}

