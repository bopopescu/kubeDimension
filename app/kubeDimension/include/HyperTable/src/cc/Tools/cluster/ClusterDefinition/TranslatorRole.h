/* -*- c++ -*-
 * Copyright (C) 2007-2015 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 3 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/// @file
/// Declarations for TranslatorRole.
/// This file contains type declarations for TranslatorRole, a class for
/// translating a role definiton statement.

#ifndef Tools_cluster_TranslatorRole_h
#define Tools_cluster_TranslatorRole_h

#include "TranslationContext.h"
#include "Translator.h"

#include <string>

namespace Hypertable { namespace ClusterDefinition {

  using namespace std;

  /// @addtogroup ClusterDefinition
  /// @{

  /// Translates a role definition.
  class TranslatorRole : public Translator {
  public:
    /// Constructor.
    /// @param fname Filename of source file
    /// @param lineno Starting offset within source file of input text
    /// @param text Input text
    TranslatorRole(const string &fname, size_t lineno, const string &text)
      : m_fname(fname), m_lineno(lineno), m_text(text) {};

    /// Translates a role definition.
    /// This function translates role definitions of the form:
    /// <pre>
    /// role: slave test[00-99]
    /// </pre>
    /// To a bash variable definition of the form:
    /// <pre>
    /// ROLE_slave="test[00-99]"
    /// </pre>
    /// As a side effect, it inserts the role name into
    /// <code>context.roles</code> and inserts the variable name representing
    /// the role (e.g. <code>ROLE_name</code>) into the
    /// <code>context.symbols</code> map, mapping it to the role value (e.g.
    /// <code>test[00-99]</code> in the above example).
    /// @param context Context object containing symbol tables
    /// @return Translated role statement.
    const string translate(TranslationContext &context) override;

  private:
    /// Source file name containing input text
    string m_fname;
    /// Starting offset within #m_fname of input text
    size_t m_lineno;
    /// Input text containing role definition statememt
    string m_text;
  };

  /// @}

}}

#endif // Tools_cluster_TranslatorRole_h
