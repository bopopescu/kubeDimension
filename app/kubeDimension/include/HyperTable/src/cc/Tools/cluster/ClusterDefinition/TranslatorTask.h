/* -*- c++ -*-
 * Copyright (C) 2007-2015 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 3 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/// @file
/// Declarations for TranslatorTask.
/// This file contains type declarations for TranslatorTask, a class for
/// translating a task definiton statement.

#ifndef Tools_cluster_TranslatorTask_h
#define Tools_cluster_TranslatorTask_h

#include "TranslationContext.h"
#include "Translator.h"

#include <string>

namespace Hypertable { namespace ClusterDefinition {

  using namespace std;

  /// @addtogroup ClusterDefinition
  /// @{

  /// Translates a task definition.
  class TranslatorTask : public Translator {
  public:
    /// Constructor.
    /// @param fname Filename of source file containing task definition
    /// @param lineno Line number within source file of task definition
    /// @param text Text of task definition
    TranslatorTask(const string &fname, size_t lineno, const string &text)
      : m_fname(fname), m_lineno(lineno), m_text(text) {};

    /// Translates a task definition.
    /// The best way to illustrate how tasks get translated is with an example.
    /// The following task definition:
    /// <pre>
    ///   task: display_hostnames roles: master,slave {
    ///     echo "before"
    ///     ssh: {
    ///       hostname
    ///     }
    ///     echo "after"
    ///   }
    /// </pre>
    /// gets translated into a bash function similar to the following:
    /// <pre>
    ///   display_hostnames () {
    ///     local _SSH_HOSTS="(${ROLE_master}) + (${ROLE_slave})"
    ///     if [ $# -gt 0 ] && [ $1 == "on" ]; then
    ///       shift
    ///       if [ $# -eq 0 ]; then
    ///         echo "Missing host specification in 'on' argument"
    ///         exit 1
    ///       else
    ///         _SSH_HOSTS="$1"
    ///         shift
    ///       fi
    ///     fi
    ///     echo "display_hostnames $@"
    ///     echo "before"
    ///     /opt/hypertable/0.9.8.1/bin/ht ssh " ${_SSH_HOSTS}" "hostname"
    ///     echo "after"
    ///   }
    /// </pre>
    /// Notice that the <code>ssh:</code> statemnts get translated to a call to
    /// the <code>ht_ssh</code> tool.  Also, the function includes logic to
    /// check for <code>on &lt;hostspec&gt;</code> initial arguments.  If found,
    /// the target hosts for <code>ssh:</code> statements becomes
    /// <code>&lt;hostspec&gt;</code> instead of the default roles specified by
    /// the <code>roles:</code> clause.
    /// @param context %Context object containing symbol tables
    /// @return translated task statement
    const string translate(TranslationContext &context) override;

  private:
    /// Source file name containing task definition
    string m_fname;
    /// Line number within #m_fname of task definition
    size_t m_lineno;
    /// Text of task definition
    string m_text;
  };

  /// @}

}}

#endif // Tools_cluster_TranslatorTask_h
