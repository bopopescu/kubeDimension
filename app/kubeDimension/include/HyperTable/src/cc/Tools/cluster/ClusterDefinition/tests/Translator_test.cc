/*
 * Copyright (C) 2007-2015 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 3 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */
#include <Common/Compat.h>

#include "../TranslationContext.h"
#include "../TranslatorRole.h"
#include "../TranslatorTask.h"

#include <Common/Error.h>
#include <Common/Logger.h>

#include <fstream>
#include <iostream>

namespace {
  const char *bad_tasks[] = {
    "# This task is just a comment",
    "# This task is missing the body\n  task:",
    "# This task is missing the end curly\n  task: roles: master {",
    "# This task is missing the roles token\n  task: { }",
    "# This task is missing roles\n  task: foo roles: { }",
    "# This task has an invalid name\n  task: 8foo { }",
    "# This task has an invalid name too\n  task: foo-bar { }",
    "# This task references an undefined role\n  task: foo roles: slave { }",
    "task: foo {\n  ssh: }",
    "task: foo {\n  ssh: bad-option=1 { hostname } }",
    "task: foo {\n  ssh: { hostname }\n  echo 'yes'\n  echo 'maybe'\n  ssh: }",
    "task: foo {\n  ssh { hostname }\n  echo 'yes'\n  echo 'maybe'\n }",
    "task: foo {\n  echo 'one'\n  ssh in-series { hostname }\n }",
    "task: foo {\n  echo 'one'\n  echo 'two'\n  ssh random-start-delay=5000 { hostname }\n }",
    "task: foo {\n  echo 'before'\n  ssh: { hostname }\n  echo 'yes'\n  echo 'maybe'\n  ssh: }",
    "# Conflict with builtin\n  task: CLUSTER_BUILTIN_display_line { }",
    "# Conflict with builtin\n  task: show_variables { }",
    (const char *)0
  };
  const char *bad_roles[] = {
    "role: foo ht-good-master\n",
    "role: foo ht-good - master\n",
    "role: foo ht-good -master\n",
    "role: foo ht-good -slave\n",
    nullptr
  };
}

using namespace Hypertable;
using namespace Hypertable::ClusterDefinition;
using namespace std;

int main(int argc, char **argv) {
  TranslationContext context;

  if (argc != 2) {
    cout << "usage: Translator_test <golden-file>" << endl;
    exit(1);
  }

  ofstream output_file;
  output_file.open("Translator_test.output");

  for (size_t i=0; bad_tasks[i]; i++) {
    try {
      TranslatorTask task("test.def", 8, bad_tasks[i]);
      string translated = task.translate(context);
      HT_FATALF("Bad task %d test failure", (int)i);
    }
    catch (Exception &e) {
      output_file << Error::get_text(e.code()) << " - " << e.what() << endl;
    }
  }

  context.roles.insert("master");
  for (size_t i=0; bad_roles[i]; i++) {
    try {
      TranslatorRole role("test.def", 8, bad_roles[i]);
      string translated = role.translate(context);
      output_file << "ok: " << translated << endl;
    }
    catch (Exception &e) {
      output_file << Error::get_text(e.code()) << " - " << e.what() << endl;
    }
  }
    
  output_file.close();

  string cmd = format("diff Translator_test.output %s", argv[1]);
  if (system(cmd.c_str()))
    exit(1);

  return 0;
}
