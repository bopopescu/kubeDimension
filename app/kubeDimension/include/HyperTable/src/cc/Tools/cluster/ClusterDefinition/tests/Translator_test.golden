HYPERTABLE syntax error - Bad task definition on line 8 of 'test.def'
HYPERTABLE syntax error - Bad task definition on line 8 of 'test.def'
HYPERTABLE syntax error - Missing terminating '}' in task definition on line 8 of 'test.def'
HYPERTABLE syntax error - Bad task definition on line 8 of 'test.def'
HYPERTABLE syntax error - Bad task definition on line 8 of 'test.def'
HYPERTABLE syntax error - Invalid task name (8foo) on line 8 of 'test.def'
HYPERTABLE syntax error - Invalid task name (foo-bar) on line 8 of 'test.def'
HYPERTABLE syntax error - Undefined role (slave) on line 8 of 'test.def'
HYPERTABLE syntax error - Invalid ssh: statement (curly brace not found) on line 9 of 'test.def'
HYPERTABLE syntax error - Invalid ssh: statement (Unknown option 'bad-option=1') on line 9 of 'test.def'
HYPERTABLE syntax error - Invalid ssh: statement (curly brace not found) on line 12 of 'test.def'
HYPERTABLE syntax error - Invalid ssh: statement on line 9 of 'test.def'
HYPERTABLE syntax error - Invalid ssh: statement on line 10 of 'test.def'
HYPERTABLE syntax error - Invalid ssh: statement on line 11 of 'test.def'
HYPERTABLE syntax error - Invalid ssh: statement (curly brace not found) on line 13 of 'test.def'
HYPERTABLE syntax error - Task name 'CLUSTER_BUILTIN_display_line' conflicts with built-in on line 8 of 'test.def'
HYPERTABLE syntax error - Task name 'show_variables' conflicts with built-in on line 8 of 'test.def'
ok: ROLE_foo="ht-good-master"

ok: ROLE_foo="ht-good - ${ROLE_master}"

ok: ROLE_foo="ht-good -${ROLE_master}"

ok: ROLE_foo="ht-good -slave"

